File: project2/js_server/routes/subway.js
import express from 'express';
const router = express.Router();

import Subway from '../models/Subway.js';

// GET /subway/getsubway?station=...
router.get('/getsubway', async (req, res) => {
  const { station } = req.query;
  if (!station) {
    return res.status(400).json({ result: false, message: 'station 쿼리 필요' });
  }
  try {
    // 역명/위치에 대해 대소문자 무시 정규식 검색
    const query = {
      $or: [
        { '역명': { $regex: station, $options: 'i' } },
        { '위치': { $regex: station, $options: 'i' } }
      ]
    };
    const docs = await Subway.find(query).lean();
    const results = docs.map(doc => ({
      stationName: doc['역명'],
      location: doc['위치'],
      openDate: doc['개통일'],
      bubjungdongCode: doc['법정동코드'] ? doc['법정동코드'].slice(0, 5) : null,
      stationRoadAddress: doc['역사도로명주소'] || ''
    }));
    return res.json({
      result: !!results.length,
      resultCount: results.length,
      results
    });
  } catch (err) {
    return res.status(500).json({ result: false, message: err.message });
  }
});

export default router;

File: project2/subway.py
import json
import httpx
from fastapi import FastAPI, Query, APIRouter, HTTPException, Depends
from typing import Dict, Optional
from pydantic import BaseModel
from pymongo.errors import PyMongoError, OperationFailure
from database import client
from datetime import datetime

class SubwayItem(BaseModel):
    """지하철역 정보"""
    stationName: str
    location: str
    openDate: str

app = FastAPI()
router = APIRouter()

@router.get('/getsubway')
async def GetSubway(
    station: str = Query(
        None,
        title="검색어",
        description="검색할 지하철역 또는 지역명을 입력하세요. (입력하지 않으면 이전에 입력한 값이 자동으로 사용)",
        min_length=0,
        required=False
    )
) -> Dict:
    try:
        mydb = client["subway"]
        subway_col = mydb["subway"]
        # 검색 쿼리 생성
        query = {
            "$or": [
                {"역명": {"$regex": station, "$options": "i"}},
                {"위치": {"$regex": station, "$options": "i"}}
            ]
        } if station else {}
        results = []
        for doc in subway_col.find(query):
            import re
            open_date = doc.get("개통일", "")
            formatted_date = ""
            if open_date:
                # 다양한 포맷 시도
                # 1. YYYY-MM-DD, YYYY.MM.DD, YYYY/MM/DD 등
                m = re.match(r"(\d{4})[-./](\d{2})[-./](\d{2})", open_date)
                if m:
                    formatted_date = f"{m.group(1)}{m.group(2)}{m.group(3)}"
                else:
                    # 2. YYYY년 MM월 DD일
                    m = re.match(r"(\d{4})년\s*(\d{1,2})월\s*(\d{1,2})일", open_date)
                    if m:
                        formatted_date = f"{m.group(1)}{int(m.group(2)):02d}{int(m.group(3)):02d}"
                    else:
                        # 3. YYYYMMDD (이미 숫자만)
                        digits = ''.join(filter(str.isdigit, open_date))
                        if len(digits) == 8:
                            formatted_date = digits
                        else:
                            # 4. YYYY-MM-DD HH:MM:SS 등
                            m = re.match(r"(\d{4})[-./](\d{2})[-./](\d{2})", open_date)
                            if m:
                                formatted_date = f"{m.group(1)}{m.group(2)}{m.group(3)}"
                            else:
                                formatted_date = ""
            # 위치(행정구역명)로 법정동코드 조회
            try:
                url = "http://localhost:8000/bubjungdong/getbubjungdongcode"
                params = {"location": doc["위치"]}
                bubjungdongCode = None
                async with httpx.AsyncClient() as async_client:
                    resp = await async_client.get(url, params=params)
                    data = resp.json()
                    print(f"법정동코드 API 응답: {data}")
                    if data.get("result"):
                        bubjungdongCode = data["bubjungdongCode"]
            except Exception as e:
                print(f"법정동코드 조회 에러: {e}")
                bubjungdongCode = None
            results.append({
                "stationName": doc["역명"],
                "location": doc["위치"],
                "openDate": formatted_date,
                "bubjungdongCode": bubjungdongCode,
                "stationRoadAddress": doc.get("역사도로명주소", "")
            })
        return {
            "result": bool(results),
            "resultCount": len(results),
            "results": results
        }
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

app.include_router(router)
